subcommand/connect-sidecar/command_test.go

```diff
@@ -0,0 +1,228 @@
+		},
+	}
+	for _, c := range cases {
+		t.Run(c.ExpErr, func(t *testing.T) {
+			configFile := filepath.Join(tmpDir, "svc.hcl")
+			err = ioutil.WriteFile(configFile, []byte(c.FileContents), 0600)
+			require.NoError(t, err)
+			defer func() { os.Remove(configFile) }()
+
+			ui := cli.NewMockUi()
+			cmd := Command{
+				UI: ui,
+			}
+
+			responseCode := cmd.Run([]string{"-service-config", configFile})
+			require.Equal(t, 1, responseCode, ui.ErrorWriter.String())
+			require.Contains(t, ui.ErrorWriter.String(), c.ExpErr)
+		})
+	}
+}
+
+// Test that we register the services.
+func TestRun_RegistersServices(t *testing.T) {
+	t.Parallel()
+	tmpDir, err := ioutil.TempDir("", "")
+	require.NoError(t, err)
+	defer func() { os.RemoveAll(tmpDir) }()
+
+	configFile := filepath.Join(tmpDir, "svc.hcl")
+	err = ioutil.WriteFile(configFile, []byte(servicesRegistration), 0600)
+	require.NoError(t, err)
+
+	a := agent.NewTestAgent(t, t.Name(), `primary_datacenter = "dc1"`)
+	defer a.Shutdown()
+
+	ui := cli.NewMockUi()
+	cmd := Command{
+		UI:           ui,
+		consulClient: a.Client(),
+	}
+
+	// Run async because we need to kill it when the test is over.
+	exitChan := runCommandAsynchronously(&cmd, []string{
+		"-http-addr", a.HTTPAddr(),
+		"-service-config", configFile,
+		"-sync-period", "100ms",
+	})
+	defer stopCommand(t, &cmd, exitChan)
+
+	timer := &retry.Timer{Timeout: 1 * time.Second, Wait: 100 * time.Millisecond}
+	retry.RunWith(timer, t, func(r *retry.R) {
+		svc, _, err := a.Client().Agent().Service("service-id", nil)
+		require.NoError(r, err)
+		require.Equal(r, 80, svc.Port)
+
+		svcProxy, _, err := a.Client().Agent().Service("service-id-sidecar-proxy", nil)
+		require.NoError(r, err)
+		require.Equal(r, 2000, svcProxy.Port)
+	})
+}
+
+// Test that we register services when the Consul agent is down at first.
+// In this test we use an http server to mimic Consul and we start it
+// after we start the command.
+func TestRun_RegistersServicesConsulDown(t *testing.T) {
+	t.Parallel()
+	tmpDir, err := ioutil.TempDir("", "")
+	require.NoError(t, err)
+	defer func() { os.RemoveAll(tmpDir) }()
+
+	configFile := filepath.Join(tmpDir, "svc.hcl")
+	err = ioutil.WriteFile(configFile, []byte(servicesRegistration), 0600)
+	require.NoError(t, err)
+
+	ui := cli.NewMockUi()
+	cmd := Command{
+		UI: ui,
+	}
+	randomPort := rand.Int()%10000 + 10000
+	// Run async because we need to kill it when the test is over.
+	exitChan := runCommandAsynchronously(&cmd, []string{
+		"-http-addr", fmt.Sprintf("127.0.0.1:%d", randomPort),
+		"-service-config", configFile,
+		"-sync-period", "100ms",
+	})
+	defer stopCommand(t, &cmd, exitChan)
+
+	// Start the HTTP server after 500ms.
+	time.Sleep(500 * time.Millisecond)
+	type APICall struct {
+		Method string
+		Path   string
+	}
+	var consulAPICalls []APICall
+	go func() {
+		err := http.ListenAndServe(fmt.Sprintf(":%d", randomPort), http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
It was because I didn't think I could configure the port for the Consul test agent and that I could only use whichever port it gave me. This wouldn't work because I need to know the port beforehand so I can start the command before the agent is actually listening on that port.

Thanks to your comment, I tried setting the port in the config and it actually worked so I've changed this to use the test agent now!